### PR TITLE
builtins.pyi: Use two type vars

### PR DESCRIPTION
**Pull request by utkarsh2102**
Fixes: #3201

### CODE REVIEW

**Comment by utkarsh2102** [comment]
> The best we can probably do here is to use a liberal sprinkling of Any.

Ack, makes sense. Thank you!

**Comment by wbolster-eiq** [comment]
thanks all, but it seems to me the fallback `@overload` is now less strict than it was before?

```
@overload
def iter(__function: Callable[[], Any], __sentinel: Any) -> Iterator[Any]: ...
```

the original was:

```
def iter(__function: Callable[[], _T], __sentinel: _T) -> Iterator[_T]: ...
```

...which seems more strict and hence more correct?

i am curious to learn why adding a special case for the (indeed common) `None` sentinel is not sufficient?

**Comment by srittau** [comment]
@wbolster-eiq  Without the overload, the following valid case is rejected:

```python
def foo() -> Union[str, int]:
    return ""
reveal_type(iter(foo, -1))
```

That said, keeping the other overload as second-to-last makes sense to keep the more precise typing for the common case. Would you care to send a PR?

**Comment by wbolster-eiq** [comment]
@srittau do i understand correctly that the complete definition should be like this?

```python
@overload
def iter(__iterable: Iterable[_T]) -> Iterator[_T]: ...

@overload
def iter(__function: Callable[[], Optional[_T]], __sentinel: None) -> Iterator[_T]: ...

@overload
def iter(__function: Callable[[], _T], __sentinel: _T) -> Iterator[_T]: ...
```

if so, yeah, i can turn that into a pull request, but right now i'm not sure i understand 100%.

**Comment by srittau** [comment]
@wbolster-eiq: We would still need the `Any` overload as last one to allow the example I gave. But correct apart from that.

File: stdlib/2and3/builtins.pyi [hunk 1]
```diff
@@ -1189,7 +1189,9 @@ else:
 @overload
 def iter(__iterable: Iterable[_T]) -> Iterator[_T]: ...
 @overload
-def iter(__function: Callable[[], _T], __sentinel: _T) -> Iterator[_T]: ...
+def iter(__function: Callable[[], Optional[_T]], __sentinel: None) -> Iterator[_T]: ...
+@overload
+def iter(__function: Callable[[], Any], __sentinel: Any) -> Iterator[Any]: ...
```
Comment (wbolster-eiq) [inline stdlib/2and3/builtins.pyi]: the `Any` is too relaxed here. the output from the callable will always be the same as the value yielded by the iterator, regardless of what the sentinel type is. i will open a pull request to make this more strict.

Comment (wbolster) [reply stdlib/2and3/builtins.pyi]: #3326

**Comment by wbolster** [comment]
i've opened #3326 as a follow-up to this pull request.

**Pull request merged**